### PR TITLE
(Backend) Ability to make Quotes Favorite

### PR DESCRIPTION
**Pull request by rizwanmustafa**

### CODE REVIEW

File: backend/api/routes/users.ts [hunk 1]
```diff
@@ -304,4 +304,37 @@ router.get(
   asyncHandler(UserController.getPersonalBests)
 );
 
+router.get(
+  "/favQuotes",
```
Comment (Bruception) [inline backend/api/routes/users.ts:308]: favoriteQuotes

**Review by Miodec — CHANGES_REQUESTED** [review]
Lets add a limit of 100 favorite quotes

**Comment by rizwanmustafa** [comment]
> Lets add a limit of 100 favorite quotes

Done

File: backend/api/routes/quotes.ts [hunk 2]
```diff
@@ -10,23 +11,23 @@ import {
   validateRequest,
 } from "../../middlewares/api-utils";
 
-const quotesRouter = Router();
+const router = Router();
```
Comment (Bruception) [inline backend/api/routes/quotes.ts:14]: Why did you change this?

File: backend/api/controllers/user.ts [hunk 3]
```diff
@@ -296,3 +296,39 @@ export async function getPersonalBests(
...
+export async function addFavoriteQuote(
+  req: MonkeyTypes.Request
+): Promise<MonkeyResponse> {
+  const { uid } = req.ctx.decodedToken;
+
+  const { language, quoteId } = req.body;
+  const addResult = await UsersDAO.addFavoriteQuote(uid, language, quoteId);
+
+  if (addResult === false) {
+    return new MonkeyResponse("Too many favorite quotes", null, 400);
```
Comment (Bruception) [inline backend/api/controllers/user.ts:319]: Prefer to throw a MonkeyError here

File: backend/api/routes/quotes.ts [hunk 4]
```diff
@@ -140,4 +141,45 @@ quotesRouter.post(
...
+  asyncHandler(UserController.getFavoriteQuotes)
+);
+
+router.post(
+  "/favorites",
+  RateLimit.quoteFavoritePost,
+  authenticateRequest(),
+  validateRequest({
+    body: {
+      language: joi.string().required(),
```
Comment (Bruception) [inline backend/api/routes/quotes.ts:157]: This schema is repeated. Consider storing it in a variable and reusing it.

File: backend/dao/user.js [hunk 5]
```diff
@@ -399,6 +399,72 @@ class UsersDAO {
...
+        user.favoriteQuotes[language] &&
+        user.favoriteQuotes[language].includes(quoteId)
+      ) {
+        return true;
+      }
+
+      let quotesLength = 0;
+
+      Object.keys(user.favoriteQuotes).forEach((lang) => {
+        quotesLength += user.favoriteQuotes[lang].length;
```
Comment (Bruception) [inline backend/dao/user.js:431]: This can be refactored. Try _.sumBy

File: backend/dao/user.js [hunk 6]
```diff
@@ -399,6 +399,72 @@ class UsersDAO {
...
+
+    if (!user) {
+      throw new MonkeyError(409, "User not found", "getFavoriteQuotes");
+    }
+
+    return user.favoriteQuotes ?? {};
+  }
+
+  static async addFavoriteQuote(uid, language, quoteId) {
+    const user = await db.collection("users").findOne({ uid });
```
Comment (Bruception) [inline backend/dao/user.js:414]: Db.collection(“users”) is invoked twice. Consider storing it in a variable.

File: backend/dao/user.js [hunk 7]
```diff
@@ -399,6 +399,72 @@ class UsersDAO {
       return user?.personalBests?.[mode];
     }
   }
+
+  static async getFavoriteQuotes(uid) {
+    const user = await db.collection("users").findOne({ uid });
+
+    if (!user) {
```
Comment (Bruception) [inline backend/dao/user.js:406]: Repetitive check. Consider making a function that checks if a user exists. Although there might be a better way…

File: backend/api/routes/quotes.ts [hunk 8]
```diff
@@ -140,4 +141,45 @@ quotesRouter.post(
...
+  "/favorites",
+  RateLimit.quoteFavoriteDelete,
+  authenticateRequest(),
+  validateRequest({
+    body: {
+      language: joi.string().required(),
+      quoteId: joi
+        .string()
+        .min(1)
+        .regex(/[0-9]+/)
```
Comment (Bruception) [inline backend/api/routes/quotes.ts:178]: Simpler Regex: \d+

File: backend/dao/user.js [hunk 9]
```diff
@@ -399,6 +399,72 @@ class UsersDAO {
...
+        return true;
+      }
+
+      let quotesLength = 0;
+
+      Object.keys(user.favoriteQuotes).forEach((lang) => {
+        quotesLength += user.favoriteQuotes[lang].length;
+      });
+
+      if (quotesLength >= 100) return false;
```
Comment (Bruception) [inline backend/dao/user.js:434]: Consider making this amount a configuration.

File: backend/dao/user.js (hunk 7 above)
Comment (rizwanmustafa) [reply backend/dao/user.js:406]: I am thinking of creating a function and using it directly below the `authenticateRequest` when we create the route. What do you say?

File: backend/constants/base-configuration.ts [hunk 10]
```diff
@@ -29,6 +29,7 @@ const BASE_CONFIGURATION: MonkeyTypes.Configuration = {
   useRedisForBotTasks: {
     enabled: false,
   },
+  maxFavoriteQuotes: 100,
```
Comment (Bruception) [inline backend/constants/base-configuration.ts:32]: Follow the pattern of the other configuration points.

File: backend/dao/user.js [hunk 11]
```diff
@@ -399,6 +399,78 @@ class UsersDAO {
...
+      {
+        $push: {
+          [`favoriteQuotes.${language}`]: quoteId,
+        },
+      }
+    );
+  }
+
+  static async removeFavoriteQuote(uid, language, quoteId) {
+    const user = await db.collection("users").findOne({ uid });
```
Comment (Bruception) [inline backend/dao/user.js:454]: Invoking this twice consider variable.

File: backend/constants/base-configuration.ts [hunk 12]
```diff
@@ -29,7 +29,9 @@ const BASE_CONFIGURATION: MonkeyTypes.Configuration = {
   useRedisForBotTasks: {
     enabled: false,
   },
-  maxFavoriteQuotes: 100,
+  favoriteQuotes: {
+    max: 100,
```
Comment (Bruception) [inline backend/constants/base-configuration.ts:33]: Max what?

File: backend/dao/user.js [hunk 13]
```diff
@@ -399,6 +399,77 @@ class UsersDAO {
       return user?.personalBests?.[mode];
     }
   }
+
+  static async getFavoriteQuotes(uid) {
+    const user = await db.collection("users").findOne({ uid });
+
+    if (!user) {
+      throw new MonkeyError(409, "User not found", "getFavoriteQuotes");
```
Comment (Bruception) [inline backend/dao/user.js:407]: Is this check needed? `user?.favoriteQuotes` should be fine

File: backend/dao/user.js [hunk 14]
```diff
@@ -399,6 +399,77 @@ class UsersDAO {
...
+    if (!user) {
+      throw new MonkeyError(409, "User does not exist", "deleteFavoriteQuote");
+    }
+
+    if (
+      !user.favoriteQuotes ||
+      !user.favoriteQuotes[language] ||
+      !user.favoriteQuotes[language].includes(quoteId)
+    ) {
+      return true;
```
Comment (Bruception) [inline backend/dao/user.js:465]: Why does this function return anything? I think making it's return type void is the way.

File: backend/dao/user.js [hunk 15]
```diff
@@ -399,6 +399,77 @@ class UsersDAO {
...
+          [`favoriteQuotes.${language}`]: quoteId,
+        },
+      }
+    );
+  }
+
+  static async removeFavoriteQuote(uid, language, quoteId) {
+    const usersCollection = await db.collection("users");
+    const user = await usersCollection.findOne({ uid });
+    if (!user) {
```
Comment (Bruception) [inline backend/dao/user.js:456]: We should wrap this check with a function that will throw an error.

File: backend/dao/user.js [hunk 16]
```diff
@@ -399,6 +399,77 @@ class UsersDAO {
...
+      if (quotesLength >= maxQuotes) {
+        throw new MonkeyError(
+          403,
+          "Too many favorite quotes",
+          "addFavoriteQuote"
+        );
+      }
+    }
+
+    return await usersCollection.updateOne(
```
Comment (Bruception) [inline backend/dao/user.js:443]: Why is this function returning anything?

File: backend/dao/user.js [hunk 17]
```diff
@@ -399,6 +399,77 @@ class UsersDAO {
...
+    const user = await db.collection("users").findOne({ uid });
+
+    if (!user) {
+      throw new MonkeyError(409, "User not found", "getFavoriteQuotes");
+    }
+
+    return user.favoriteQuotes ?? {};
+  }
+
+  static async addFavoriteQuote(uid, language, quoteId, maxQuotes) {
```
Comment (Bruception) [inline backend/dao/user.js:413]: Fix the weird return types of this function(shouldn't return anything)

File: backend/dao/user.js (hunk 15 above)
Comment (rizwanmustafa) [reply backend/dao/user.js:456]: Will do that in a separate PR

File: backend/api/routes/users.ts [hunk 18]
```diff
@@ -304,4 +307,37 @@ router.get(
   asyncHandler(UserController.getPersonalBests)
 );
 
+router.get(
+  "/favoriteQuotes",
```
Comment (Bruception) [inline backend/api/routes/users.ts:311]: We should move this back to the quotes routes. Despite storing them in the users collection.

File: backend/dao/user.ts [hunk 19]
```diff
@@ -534,3 +534,85 @@ export async function getPersonalBests(
 
   return user?.personalBests?.[mode];
 }
+
+export async function getFavoriteQuotes(
+  uid
+): Promise<MonkeyTypes.User["favoriteQuotes"]> {
+  const user = await db.collection<MonkeyTypes.User>("users").findOne({ uid });
+
+  if (!user) {
```
Comment (Bruception) [inline backend/dao/user.ts:543]: Still aren’t wrapping this in a function

File: backend/api/routes/users.ts (hunk 18 above)
Comment (rizwanmustafa) [reply backend/api/routes/users.ts:311]: Miodec asked me to have this is in the users routes

File: backend/dao/user.ts (hunk 19 above)
Comment (Miodec) [reply backend/dao/user.ts:543]: Separate pr

**Review by Miodec — APPROVED** [review]
Good to me

**Pull request merged**